pkg/crd/markers/validation.go

```diff
@@ -635,3 +659,44 @@ func (m XValidation) ApplyToSchema(schema *apiext.JSONSchemaProps) error {
 	})
 	return nil
 }
+
+func (fields AtMostOneOf) ApplyToSchema(schema *apiext.JSONSchemaProps) error {
+	if len(fields) == 0 {
+		return nil
+	}
+	rule := fieldsToOneOfCelRuleStr(fields)
+	xvalidation := XValidation{
+		Rule:    fmt.Sprintf("%s <= 1", rule),
+		Message: fmt.Sprintf("at most one of the fields in %v can be set", fields),
+	}
+	return xvalidation.ApplyToSchema(schema)
+}
+
+func (fields ExactlyOneOf) ApplyToSchema(schema *apiext.JSONSchemaProps) error {
+	if len(fields) == 0 {
+		return nil
+	}
+	rule := fieldsToOneOfCelRuleStr(fields)
+	xvalidation := XValidation{
+		Rule:    fmt.Sprintf("%s == 1", rule),
+		Message: fmt.Sprintf("exactly one of the fields in %v must be set", fields),
+	}
+	return xvalidation.ApplyToSchema(schema)
+}
+
+// fieldsToOneOfCelRuleStr converts a slice of field names to a string representation
+// [has(self.field1),has(self.field1),...].filter(x, x == true).size()
+func fieldsToOneOfCelRuleStr(fields []string) string {
+	var list strings.Builder
+	list.WriteString("[")
+	for i, f := range fields {
+		if i > 0 {
+			list.WriteString(",")
+		}
+		list.WriteString("has(self.")
```

Comment:
Out of interest, I wonder if you could set a second child field using this 🤔 Maybe we should add a test to see if that works (and then prevent us breaking it later if it does)

Reply:
dup of https://github.com/kubernetes-sigs/controller-tools/pull/1212#discussion_r2145499984